src/engine/CountAvailablePredicates.cpp

```diff
@@ -97,40 +112,147 @@ void CountAvailablePredicates::computeResult(ResultTable* result) const {
+      } else if (subresult->_nofColumns == 2) {
+        CountAvailablePredicates::computePatternTrick<array<Id, 2>>(
+            static_cast<vector<array<Id, 2>>*>(subresult->_fixedSizeData),
+            static_cast<vector<array<Id, 2>>*>(result->_fixedSizeData),
+            hasPattern, hasPredicate, patterns, _subjectColumnIndex);
+      } else if (subresult->_nofColumns == 3) {
+        CountAvailablePredicates::computePatternTrick<array<Id, 3>>(
+            static_cast<vector<array<Id, 3>>*>(subresult->_fixedSizeData),
+            static_cast<vector<array<Id, 2>>*>(result->_fixedSizeData),
+            hasPattern, hasPredicate, patterns, _subjectColumnIndex);
+      } else if (subresult->_nofColumns == 4) {
+        CountAvailablePredicates::computePatternTrick<array<Id, 4>>(
+            static_cast<vector<array<Id, 4>>*>(subresult->_fixedSizeData),
+            static_cast<vector<array<Id, 2>>*>(result->_fixedSizeData),
+            hasPattern, hasPredicate, patterns, _subjectColumnIndex);
+      } else if (subresult->_nofColumns == 5) {
+        CountAvailablePredicates::computePatternTrick<array<Id, 5>>(
+            static_cast<vector<array<Id, 5>>*>(subresult->_fixedSizeData),
+            static_cast<vector<array<Id, 2>>*>(result->_fixedSizeData),
+            hasPattern, hasPredicate, patterns, _subjectColumnIndex);
+      }
     }
   }
   result->finish();
 }
+
+void CountAvailablePredicates::computePatternTrickAllEntities(
+    vector<array<Id, 2>>* result, const vector<PatternID>& hasPattern,
+    const CompactStringVector<Id, Id>& hasPredicate,
+    const CompactStringVector<size_t, Id>& patterns) {
+  ad_utility::HashMap<Id, size_t> predicateCounts;
+  ad_utility::HashMap<size_t, size_t> patternCounts;
+
+  size_t maxId = std::max(hasPattern.size(), hasPredicate.size());
+  for (size_t i = 0; i < maxId; i++) {
+    if (i < hasPattern.size() && hasPattern[i] != NO_PATTERN) {
+      patternCounts[hasPattern[i]]++;
+    } else if (i < hasPredicate.size()) {
+      size_t numPredicates;
+      Id* predicateData;
+      std::tie(predicateData, numPredicates) = hasPredicate[i];
+      if (numPredicates > 0) {
+        for (size_t i = 0; i < numPredicates; i++) {
+          auto it = predicateCounts.find(predicateData[i]);
+          if (it == predicateCounts.end()) {
+            predicateCounts[predicateData[i]] = 1;
+          } else {
+            it->second++;
+          }
+        }
+      }
+    }
+  }
+
+  for (const auto& it : patternCounts) {
+    std::pair<Id*, size_t> pattern = patterns[it.first];
+    for (size_t i = 0; i < pattern.second; i++) {
+      predicateCounts[pattern.first[i]] += it.second;
+    }
+  }
+  result->reserve(predicateCounts.size());
+  for (const auto& it : predicateCounts) {
+    result->push_back(array<Id, 2>{it.first, static_cast<Id>(it.second)});
+  }
+}
+
+template <typename A>
+void CountAvailablePredicates::computePatternTrick(
+    const vector<A>* input, vector<array<Id, 2>>* result,
```

Comment:
Can we make `input` a `const vector<A>&` or is the use of pointers a convention for inputs in the `compute*` methods?

Reply:
Never mind, I see this is called with `result->_fixedSizeData`